src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Win32.cs

```diff
@@ -242,48 +242,52 @@ private static void HandleLastWin32Error()
 
     internal static class NtProcessInfoHelper
     {
-        // Cache a single buffer for use in GetProcessInfos().
-        private static long[]? CachedBuffer;
-
         // Use a smaller buffer size on debug to ensure we hit the retry path.
 #if DEBUG
-        private const int DefaultCachedBufferSize = 1024;
+        private const int DefaultCachedBufferSize = 8 * 1024;
 #else
-        private const int DefaultCachedBufferSize = 128 * 1024;
+        private const int DefaultCachedBufferSize = 1024 * 1024;
 #endif
+        // use DefaultCachedBufferSize to ensure we test the unmanaged memory code path path at least in Debug builds
+        private const int RentSizeLimit = DefaultCachedBufferSize + 1;
+
+        private static int MostRecentSize = DefaultCachedBufferSize;
 
         internal static ProcessInfo[] GetProcessInfos(Predicate<int>? processIdFilter = null)
         {
-            ProcessInfo[] processInfos;
-
             // Start with the default buffer size.
-            int bufferSize = DefaultCachedBufferSize;
-
-            // Get the cached buffer.
-            long[]? buffer = Interlocked.Exchange(ref CachedBuffer, null);
+            int bufferSize = MostRecentSize;
 
-            try
+            while (true)
             {
-                while (true)
-                {
-                    if (buffer == null)
-                    {
-                        // Allocate buffer of longs since some platforms require the buffer to be 64-bit aligned.
-                        buffer = new long[(bufferSize + 7) / 8];
-                    }
+                (byte[]? managed, IntPtr unmanaged) = bufferSize < RentSizeLimit
```

Comment:
Is using two different ways to obtain the buffer helping performance in any way?

Reply:
FWIW, we are using `Marshal.AllocHGlobal` instead of array pool for various reasons in number of other places that are more mainstream than this API. For example, `src\libraries\System.IO.FileSystem\src\System\IO\Enumeration\FileSystemEnumerator.Windows.cs` uses `Marshal.AllocHGlobal` instead of ArrayPool.

Reply:
I am going to get the numbers for `Marshal.AllocHGlobal` only approach and get back to you

Reply:
> Is using two different ways to obtain the buffer helping performance in any way?

I've checked and when using `Marshal.AllocHGlobal` only the microbenchmarks perf falls by 10%:

 |             Method |        Job |           Toolchain |     Mean | Ratio |
 |------------------- |----------- |-------------------- |---------:|------:|
 |       GetProcesses | Job-XBZISW |  \after\CoreRun.exe | 3.543 ms |  1.10 |
 |       GetProcesses | Job-YJQLFD | \before\CoreRun.exe | 3.226 ms |  1.00 |
 |                    |            |                     |          |       |
 | GetProcessesByName | Job-XBZISW |  \after\CoreRun.exe | 3.572 ms |  1.12 |
 | GetProcessesByName | Job-YJQLFD | \before\CoreRun.exe | 3.201 ms |  1.00 |


When it comes to the sample ASP.NET app provided above it's much worse (RPS / 4):

| load                   |  before |     mix |unmanaged |
| ---------------------- | ------- | ------- | -------- |
| CPU Usage (%)          | 10      | 12      |  3       |
| Cores usage (%)        | 118     | 142     |  37      |
| Working Set (MB)       | 49      | 49      |  48      |
| Build Time (ms)        | 5,877   | 5,898   |  5,966   |
| Start Time (ms)        | 0       | 0       |  0       |
| Published Size (KB)    | 76,401  | 76,401  |  76,401  |
| First Request (ms)     | 103     | 106     |  111     |
| Requests/sec           | 13,499  | 17,852  |  4,024   |
| Requests               | 203,683 | 269,349 |  60,718  |
| Mean latency (ms)      | 38.06   | 28.65   |  126.36  |
| Max latency (ms)       | 240.17  | 168.62  |  294.71  |
| Bad responses          | 0       | 0       |  0       |
| Socket errors          | 0       | 0       |  0       |
| Read throughput (MB/s) | 1.18    | 1.57    |  0.35    |
| Latency 50th (ms)      | 38.57   | 26.50   |  122.76  |
| Latency 75th (ms)      | 46.44   | 36.76   |  163.60  |
| Latency 90th (ms)      | 57.80   | 40.75   |  183.48  |
| Latency 99th (ms)      | 84.34   | 57.51   |  209.58  |
                                              
@jkotas so I believe that it's worth keeping the current proposed soluton (`ArrayPool` for `<= 1MB` and `Marshal.AllocHGlobal` for the rest)

FWIW the source  code;

<details>

```cs
        internal static ProcessInfo[] GetProcessInfos(Predicate<int>? processIdFilter = null)
        {
            // Start with the default buffer size.
            int bufferSize = MostRecentSize;

            while (true)
            {
                IntPtr unmanaged = Marshal.AllocHGlobal(bufferSize);

                try
                {
                    unsafe
                    {
                        Span<byte> buffer = new Span<byte>(unmanaged.ToPointer(), bufferSize);

                        uint requiredSize = 0;
                        uint status = Interop.NtDll.NtQuerySystemInformation(
                            Interop.NtDll.SystemProcessInformation,
                            unmanaged.ToPointer(),
                            (uint)buffer.Length,
                            &requiredSize);

                        if (status != Interop.NtDll.STATUS_INFO_LENGTH_MISMATCH)
                        {
                            // see definition of NT_SUCCESS(Status) in SDK
                            if ((int)status < 0)
                            {
                                throw new InvalidOperationException(SR.CouldntGetProcessInfos, new Win32Exception((int)status));
                            }

                            MostRecentSize = buffer.Length;
                            // Parse the data block to get process information
                            return GetProcessInfos(buffer, processIdFilter);
                        }

                        bufferSize = GetNewBufferSize(bufferSize, (int)requiredSize);
                    }
                }
                finally
                {
                    Marshal.FreeHGlobal(unmanaged);
                }
            }
        }
```

</details>

Reply:
> When it comes to the sample ASP.NET app provided above it's much worse (RPS / 4):

Do you have explanation for why it is worse?

The CPU usage for the managed case is 10% and for the unmanaged case 3%. It suggests that there is a lot of lock contention in both cases, and the lock contention has worse effect on throughput for the unmanaged case.

This is very expensive API to call. I do not think we need to be optimizing this API for being called on all threads in tight loop. I think we should assume that the API is called once in a while (e.g. once per second) and we should be minimizing negative effects of this API on the rest of the workload.

Reply:
> FWIW the source code

I assumed that all numbers that you are sharing are for the public entrypoint. Measuring performance of the internal methods can give us additional insights, but the performance of the public entrypoints is what actually matters.

Reply:
> I assumed that all numbers that you are sharing are for the public entrypoint.

And you were right. I've shared the source code because I decided not to push it. I just wanted to be transparent about the implementation

Reply:
The benchmarks that I am using:

https://github.com/dotnet/performance/blob/4f408efd7b94a4b5875a8f79e6f6c277a1ac1618/src/benchmarks/micro/libraries/System.Diagnostics/Perf_Process.cs#L25-L41

Reply:
> you believe that using ArrayPool here is just not worth the complexity?

Yes, I believe that using ArrayPool here is not worth the complexity.

Reply:
> The CPU usage for the managed case is 10% and for the unmanaged case 3%. 

Please excuse me for the confusion... the ASP.NET benchmarking infra has the concept of load (what sends the web requests) and application (what provides web responses). Here the mentioned CPU usage was for load (a C++ app called `wrk2` that sends the web requests)

The application has close to 100% CPU usage for all the cases. Sample full results:

| application           |                         |
| --------------------- | ----------------------- |
| CPU Usage (%)         | 97                      |
| Cores usage (%)       | 2,703                   |
| Working Set (MB)      | 1,000                   |
| Build Time (ms)       | 11,961                  |
| Start Time (ms)       | 357                     |
| Published Size (KB)   | 84,768                  |
| .NET Core SDK Version | 6.0.100-alpha.1.20568.5 |


| load                   |         |
| ---------------------- | ------- |
| CPU Usage (%)          | 3       |
| Cores usage (%)        | 79      |
| Working Set (MB)       | 48      |
| Build Time (ms)        | 5,125   |
| Start Time (ms)        | 0       |
| Published Size (KB)    | 76,401  |
| .NET Core SDK Version  | 3.1.404 |
| First Request (ms)     | 107     |
| Requests/sec           | 12,599  |
| Requests               | 190,181 |
| Mean latency (ms)      | 39.82   |
| Max latency (ms)       | 260.29  |
| Bad responses          | 0       |
| Socket errors          | 0       |
| Read throughput (MB/s) | 1.11    |
| Latency 50th (ms)      | 36.26   |
| Latency 75th (ms)      | 44.25   |
| Latency 90th (ms)      | 55.47   |
| Latency 99th (ms)      | 88.49   |

Reply:
> Yes, I believe that using ArrayPool here is not worth the complexity.

It's just a few lines of code that provide better performance for the most common case. I believe that it's worth it.

Reply:
Another reason for not doing it is that it is introducing performance cliff. Performance cliffs are common source of support incidents.

Reply:
I have done a quick test to compare performance of allocating 1MB buffer from ArrayPool vs. Marshal.AllocHGlobal
```
using System;
using System.Buffers;
using System.Diagnostics;

int size = 1_000_000;
for (; ; )
{
    Stopwatch sw = new Stopwatch();
    sw.Start();
    for (int i = 0; i < 10000; i++)
    {
#if true
        byte[] b = ArrayPool<byte>.Shared.Rent(size);
        b.AsSpan().Clear();
        ArrayPool<byte>.Shared.Return(b);
#else
        IntPtr p = Marshal.AllocHGlobal(size);
        new Span<byte>((void*)p, size).Clear();
        Marshal.FreeHGlobal(p);
#endif
    }
    sw.Stop();
    Console.WriteLine(sw.ElapsedMilliseconds);
}
```
I do not see any performance difference between the two.